plugins/ccgx/fu-ccgx-hpi.c

```diff
@@ -0,0 +1,1622 @@
+}
+
+/**
+ * fu_ccgx_hpi_validate_fw:
+ * @device: #FuDevice
+ * @handle: HPI handle
+ * @fw_index firmware index (FW1 or FW2)
+ * @error: a #GError or %NULL
+ *
+ * Validate firmware written
+ *
+ * Returns: %TRUE for success
+*/
+static gboolean
+fu_ccgx_hpi_validate_fw (FuDevice *device,CyHPIHandle *handle, guint8 fw_index, GError **error)
+{
+	g_return_val_if_fail (handle != NULL, FALSE);
+	if (!fu_ccgx_hpi_reg_write(device, &handle->i2c_handle, handle->hpi_addr_byte, CY_PD_REG_VALIDATE_FW_ADDR, &fw_index, 1, error)) {
+		g_prefix_error (error, "validate fw error: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+/**
+ * fu_ccgx_hpi_reset_device:
+ * @device: #FuDevice
+ * @handle: HPI handle
+ * @error: a #GError or %NULL
+ *
+ * Reset device
+ *
+ * Returns: %TRUE for success
+*/
+static gboolean
+fu_ccgx_hpi_reset_device (FuDevice *device, CyHPIHandle *handle, GError **error)
+{
+	guint8 reset_cmd[2];
+	reset_cmd[0] = CY_PD_DEVICE_RESET_CMD_SIG;
+	reset_cmd[1] = CY_PD_REG_RESET_DEVICE_CMD;
+
+	g_return_val_if_fail (handle != NULL, FALSE);
+	if (!fu_ccgx_hpi_reg_write_no_resp(device, &handle->i2c_handle, handle->hpi_addr_byte, CY_PD_REG_RESET_ADDR, reset_cmd, 2, error)) {
+		g_prefix_error (error, "reset device error: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+/**
+ * fu_ccgx_hpi_get_event:
+ * @device: #FuDevice
+ * @handle: HPI handle
+ * @reg_section register section type. see HPIRegSection
+ * @event[out] event response from device
+ * @io_timeout timeout to wait event
+ * @error: a #GError or %NULL
+ *
+ * Get event from device within specificed timeout
+ *
+ * Returns: %TRUE for success
+*/
+static gboolean
+fu_ccgx_hpi_get_event (FuDevice *device, CyHPIHandle *handle, HPIRegSection reg_section, CyPDResp*	event, guint32 io_timeout, GError **error )
+{
+	HPIEvent event_array[HPI_REG_SECTION_ALL + 1] = {0};
+	guint32 timeout_ms = io_timeout;
+
+	g_return_val_if_fail (handle != NULL, FALSE);
+	g_return_val_if_fail (event != NULL, FALSE);
+	if (fu_ccgx_hpi_wait_for_event(device,handle,reg_section,event_array,timeout_ms,error)) {
+		*event = event_array[reg_section].event_code;
+		return TRUE;
+	}
+	return FALSE;
+}
+
+/**
+ * fu_ccgx_hpi_clear_all_event:
+ * @device: #FuDevice
+ * @handle: HPI handle
+ * @io_timeout timeout to clear event
+ * @error: a #GError or %NULL
+ *
+ * Clear all event from device
+ *
+*/
+static void
+fu_ccgx_hpi_clear_all_event (FuDevice *device,
+			     CyHPIHandle *handle,
+			     guint32 io_timeout,
+			     GError **error)
+{
+	g_autoptr(GError) error_local = NULL;
+	HPIEvent event_array[HPI_REG_SECTION_ALL + 1] = {0};
+	guint32 timeout_ms = io_timeout;
+	guint8 section;
+
+	g_return_if_fail (handle != NULL);
+	if (timeout_ms == 0) {
+		fu_ccgx_hpi_app_read_intr_reg (device, handle, HPI_REG_SECTION_ALL, event_array, &error_local);
+	} else {
+		for (section = 0; section < handle->num_of_ports; section++) {
+			fu_ccgx_hpi_wait_for_event (device, handle, section, event_array, timeout_ms, &error_local);
+		}
```

Comment:
The error handling is all wonky here. Are you printing error_local, or trying to set `error`?

Reply:
I just want to clear all events even though there is no event and error and just want to print error_local here.
Could you check my code snippet below?
 - fu_ccgx_hpi_clear_all_event -> fu_ccgx_hpi_clear_all_events 
 - remove GError in parameters
 - add error print

static void
fu_ccgx_hpi_clear_all_events (FuDevice *device, CyHPIHandle *handle, guint32 io_timeout)
{
        g_autoptr(GError) error_local = NULL;
	HPIEvent event_array[HPI_REG_SECTION_ALL + 1] = {0};
	guint32 timeout_ms = io_timeout;
	guint8 section;
        gint8 event_count;

	g_return_if_fail (handle != NULL);
	if (timeout_ms == 0) {
		event_count = fu_ccgx_hpi_app_read_intr_reg (device, 
                              handle, HPI_REG_SECTION_ALL, event_array, &error_local);
                if (event_count < 0) 
                        g_warning ("get event error:%s",error_local->message);
                else if (event_count == 0) 
                        g_warning ("get event error no event");
	} else {
		for (section = 0; section < HPI_REG_SECTION_ALL; section++) {
			event_count = fu_ccgx_hpi_wait_for_event (device, 
                                      handle, section, event_array, timeout_ms, &error_local);
                        if (event_count < 0) 
                                g_warning ("get event error:%s",error_local->message);
                        else if (event_count == 0) 
                                g_warning ("get event error no event");
                }
	}
}

Reply:
I still think fu_ccgx_hpi_clear_all_events() should use a GError() and return gboolean - then the caller can do the appropriate action if it fails.

Reply:
OK. I will make fu_ccgx_hpi_clear_all_events get process GError and get return value when it has error.

Reply:
Can you please push to this branch even when you've got a few fixes (it doesn't matter how many commits you do) and then we can move this forward with the next chunk.

Reply:
I fixed it.